### PR TITLE
Tracking extension fix for 2.11+

### LINKED ISSUE #8447: CKAN 2.11 - Error on log
## Describe the bug
Every page visited, with "tracking" plugin active, generate this error on log:
2024-09-19 19:46:54,087 ERROR [ckan.config.middleware.flask_app] Exception on /_tracking [POST]
Traceback (most recent call last):
  File "/usr/lib/ckan/default/lib/python3.10/site-packages/flask/app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/lib/ckan/default/lib/python3.10/site-packages/flask/app.py", line 883, in full_dispatch_request
    return self.finalize_request(rv)
  File "/usr/lib/ckan/default/lib/python3.10/site-packages/flask/app.py", line 904, in finalize_request
    response = self.process_response(response)
  File "/usr/lib/ckan/default/lib/python3.10/site-packages/flask/app.py", line 1281, in process_response
    response = self.ensure_sync(func)(response)
  File "/usr/lib/ckan/default/src/ckan/ckanext/tracking/middleware.py", line 42, in track_request
    with cast(sa.engine.Engine, engine).begin() as conn:
AttributeError: 'NoneType' object has no attribute 'begin'
2024-09-19 19:46:54,088 INFO  [ckan.config.middleware.flask_app] 500 Internal Server Error: The server encountered an internal error and was unable to complete your request. Either the server is overloaded or there is an error in the application.
2024-09-19 19:46:54,630 ERROR [ckan.config.middleware.flask_app] Request finalizing failed with an error while handling an error
Traceback (most recent call last):
  File "/usr/lib/ckan/default/lib/python3.10/site-packages/flask/app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
  File "/usr/lib/ckan/default/lib/python3.10/site-packages/flask/app.py", line 883, in full_dispatch_request
    return self.finalize_request(rv)
  File "/usr/lib/ckan/default/lib/python3.10/site-packages/flask/app.py", line 904, in finalize_request
    response = self.process_response(response)
  File "/usr/lib/ckan/default/lib/python3.10/site-packages/flask/app.py", line 1281, in process_response
    response = self.ensure_sync(func)(response)
  File "/usr/lib/ckan/default/src/ckan/ckanext/tracking/middleware.py", line 42, in track_request
    with cast(sa.engine.Engine, engine).begin() as conn:
AttributeError: 'NoneType' object has no attribute 'begin'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/ckan/default/lib/python3.10/site-packages/flask/app.py", line 904, in finalize_request
    response = self.process_response(response)
  File "/usr/lib/ckan/default/lib/python3.10/site-packages/flask/app.py", line 1281, in process_response
    response = self.ensure_sync(func)(response)
  File "/usr/lib/ckan/default/src/ckan/ckanext/tracking/middleware.py", line 22, in track_request
    k, v = part.split('=')
ValueError: not enough values to unpack (expected 2, got 1)

### Steps to reproduce
Activate "tracking" plugin and visit some pages on the ckan site.

### PR DESCRIPTION
**Pull request by tino097**
Fixes #8447 and #8431

### CODE REVIEW

File: ckanext/tracking/middleware.py [hunk 1]
```diff
@@ -35,15 +34,13 @@ def track_request(response: Response) -> Response:
...
-                "type": data.get("type"),
-            })
+        try:
+            TrackingRaw.create(
+                user_key=key,
+                url=data.get("url"),
+                tracking_type=data.get("type")
+            )
+        except Exception as e:
+            raise e
```
Comment (avdata99) [inline ckanext/tracking/middleware.py:44]: It's probably non required to track the exception just to raise it. 
Maybe you want to do something else (or maybe just drop the try-catch here)

File: ckanext/tracking/model.py [hunk 2]
```diff
@@ -17,53 +17,108 @@
...
+Index('tracking_raw_url', TrackingRaw.url)
+Index('tracking_raw_access_timestamp', 'access_timestamp')
+
+
+class TrackingSummary(domain_object.DomainObject, BaseModel):  # type: ignore
+
+    __tablename__ = 'tracking_summary'
+    # id = Column(
+    #     "id", types.UnicodeText, primary_key=True, default=_types.make_uuid
+    # )
```
Comment (avdata99) [inline ckanext/tracking/model.py:74]:
```suggestion
```
This is probably non-required and it could be deleted

**Comment by avdata99** [comment]
At first glance this PR looks good.
It is required that someone with more knowledge about SQLAlchemy reviews this. 
This looks like related to #7583

File: ckanext/tracking/middleware.py (hunk 1 above)
Comment (tino097) [reply ckanext/tracking/middleware.py:44]: I think its good to log the error but i forgot to add it and remove the `raise e`

**Comment by amercader** [comment]
@pdelboca I marked this for backporting for 2.11 but please check if the model changes would be compatible and if there are upgrade steps that need to be documented, thanks!

**Comment by pdelboca** [comment]
@tino097 the PR is missing the `alembic.ini` file that needs to be present in  `ckanext/tracking/migration/tracking` folder.

Currently, running a `ckan db upgrade -p tracking` will trigger and error:
```
    alembic_upgrade(self.alembic_config, version)
  File "/home/pdelboca/Repos/ckan/.venv/lib/python3.11/site-packages/alembic/command.py", line 386, in upgrade
    script = ScriptDirectory.from_config(config)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/pdelboca/Repos/ckan/.venv/lib/python3.11/site-packages/alembic/script/base.py", line 170, in from_config
    raise util.CommandError(
alembic.util.exc.CommandError: No 'script_location' key found in configuration.

```

We also need to document a new command: `ckan db upgrade -p tracking`. Even when it is not strictly necessary to for the extension to work we should add it as a installation step to avoid problems in the future (and to apply the changes proposed in this PR ofc)

@amercader codewise I think this will not fail if we backported it to `2.11` however I do think that a proper way to do it will be to add this in a small patch release and well documented (since now officially `tracking` requires a db init command).

Maybe it is time to think about CKAN 2.11.1?

**Comment by amercader** [comment]
> 

> @amercader codewise I think this will not fail if we backported it to 2.11 however I do think that a proper way to do it will be to add this in a small patch release and well documented (since now officially tracking requires a db init command).

yes, of course. Anything backported goes into a patch release and needs to be properly documented

> Maybe it is time to think about CKAN 2.11.1?

Yes it's time,already working on it, sorting out the requirements upgrades etc

**Comment by tino097** [comment]
Regarding the documenting the changes, it should be a different PR i believe @pdelboca

**Comment by amercader** [comment]
@tino097 If it's just documenting the new upgrade command let's do it in this pull request so we can backport it all in one go, thanks!

**Comment by tino097** [comment]
you got me unprepared for this :) 

i will add and push asap

**Comment by tino097** [comment]
@pdelboca @amercader update the docs referencing that we need to run the migrations 
Please let me know if i need to change/add  something

**Comment by tino097** [comment]
> alembic_upgrade(self.alembic_config, version)
>   File "/home/pdelboca/Repos/ckan/.venv/lib/python3.11/site-packages/alembic/command.py", line 386, in upgrade
>     script = ScriptDirectory.from_config(config)
>              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>   File "/home/pdelboca/Repos/ckan/.venv/lib/python3.11/site-packages/alembic/script/base.py", line 170, in from_config
>     raise util.CommandError(
> alembic.util.exc.CommandError: No 'script_location' key found in configuration.

This happens for activity too, i think we need to include `alembic.ini` into setup.cfg too

**Comment by pdelboca** [comment]
Thanks @tino097 !

**Comment by BENETNATH** [comment]
Hello,
I've just installed CKAN 2.11 on a fresh ubuntu through package .
Everything works except this tracking issue.
What would be the best option to fix the install ?
thank

**Comment by pdelboca** [comment]
Hello @BENETNATH !

The Backport has been merged and will be added in the next minor version of CKAN `2.11.1` that will be released in the upcoming weeks.

If you want to fix it locally before it, you could implement the patch that we have just merged or update the local `dev-v2.11` branch.

**Pull request merged**